### ISSUE TITLE
ManagedLedger should streamline the read requests

### ISSUE THREAD

**Issue by sijie** [apache/pulsar#3804]
*Problems*

Currently managed ledger read entries in a very large batch requests - 100 entries by default. This is an inefficient approach. We should streamline the read requests like what dlog is doing.

**Comment by sijie** [comment]
I have seen in a production deployment the consumer can never catch up if bookie's avg read latency is 10+ms bad due to the disk and other workloads running on same machine. the problem can potentially be mitigated if we can have a proper readahead mechanism in managed like what dlog is doing.

**Comment by eolivelli** [comment]
@sijie
I agree that pushing that mechanism to the low level API will be useful

**Comment by vicaya** [comment]
This is a also blocking issue for practical use of tiered storage as historical retention, since replay from tiered storage (at least s3) is too slow.

It'd be great if number of read-ahead threads and/or outstanding requests is tunable, as many blob/object store scale throughput proportional to number of connections.

**Comment by sijie** [comment]
@vicaya thank you for your feedback. @MarvinCai are you willing to give it a try?

**Comment by MarvinCai** [comment]
@sijie sorry just saw the replies, how about I start with doc with problem statement and try propose a solution. If everything looks good then we can proceed from there.

**Comment by sijie** [comment]
@MarvinCai Are you working on this issue already?

**Comment by MarvinCai** [comment]
@sijie  I was new to BK code base and was reading the some LeadferHandler and DL codes tog figure out what should be change, have a simple [doc](https://docs.google.com/document/d/1_GgFugybNQXxxvbZnPYw6PmU30NSFogEDrtYqF68f0w/edit?usp=sharing) about what I think may need to change.
I haven't start writing any code yet. If someone with more experience with BK plan to work on it then I'm fine. Else I'm also glad to help.

**Comment by vicaya** [comment]
Is there any progress on this issue? Whenever people ask me about why we don't use tiered storage, I have to point them to this issue for why it's too slow for us (readers cannot read fast enough from tiered storage and backlogs would build up even at moderate throughput (<1000msgs/s with small messages (<100 bytes), which is <100KB/s).

**Comment by eolivelli** [comment]
@nicoloboschi you could be interested in working on a fix for this issue if @MarvinCai doesn't have time to work on this topic

**Comment by michaeljmarshall** [comment]
@MarvinCai, @sijie, @eolivelli - It looks like the feature to improve read throughput for blob storage is still outstanding. I'm very interested in helping to contribute this feature. Do you think it deserves its own issue to start a dedicated discussion for offloading?

**Comment by eolivelli** [comment]
@michaeljmarshall you can start a discussion with a proposal.
you can start on dev@pulsar in order to have more people in the discussion.

**Comment by michaeljmarshall** [comment]
@eolivelli - thanks, I'll do that.
